File: aoc2022/kotlin/src/main/kotlin/Day04.kt
import util.AdventOfCode

object Day04 : AdventOfCode() {

    private fun IntRange.overlapsCompletely(b: IntRange) =
        (this.first <= b.first && this.last >= b.last) || (this.first >= b.first && this.last <= b.last)

    private fun IntRange.overlapsAtAll(b: IntRange): Boolean =
        this.first in b || this.last in b || b.first in this || b.last in this

    private fun Sequence<String>.parse() = map { it.parseLine() }

    private fun String.parseLine() =
        this.split(",").map { it.split('-').let { (start, end) -> start.toInt()..end.toInt() } }

    fun part1(input: Sequence<String>) =
        input.parse().count { (a, b) -> a.overlapsCompletely(b) }

    fun part2(input: Sequence<String>) =
        input.parse().count { (a, b) -> a.overlapsAtAll(b) }

    @JvmStatic
    fun main(args: Array<String>) {
        println("Day 4")
        println("\tPart 1: ${part1(inputFileLineSequence)}")
        println("\tPart 2: ${part2(inputFileLineSequence)}")
    }
}

File: aoc2022/fsharp/Days/Day04.fs
module Day04

let overlapsCompletely ((firstA: int, lastA: int): int * int) ((firstB: int, lastB: int): int * int) =
    (firstA <= firstB && lastA >= lastB) || (firstA >= firstB && lastA <= lastB)

let overlaps (firstA: int, lastA: int) (firstB: int, lastB: int) =
    (firstA >= firstB && firstA <= lastB) || (firstB >= firstA && firstB <= lastA)

let parseTuple (s: string) =
    let value = s.Split([| '-' |]) |> Array.map int
    (value[0], value[1])

let parseLine (s: string) =
    let tuples = s.Split([| ',' |]) |> Array.map parseTuple
    (tuples[0], tuples[1])

let parse = Seq.map parseLine

let part1 (input: string seq) =
    input
    |> parse
    |> Seq.filter (fun (a, b) -> a |> overlapsCompletely <| b)
    |> Seq.length

let part2 (input: string seq) =
    input |> parse |> Seq.filter (fun (a, b) -> a |> overlaps <| b) |> Seq.length
